### PR TITLE
[ENT-685] Improve landing page UX for when item is unenrollable.

### PR DESCRIPTION
**Pull request by UmanShahzad**
**Description:** This PR improves the landing page UX for when the course or program is unenrollable. How that's determined is given by the function `is_course_run_enrollable` being run on the course(s) -- see the docstring or the JIRA ticket.

**JIRA:** [ENT-685](https://openedx.atlassian.net/browse/ENT-685)

**Merge deadline:** End of sprint

**Testing instructions:**

1. Configure a course run to have the end date before now, or use any of the other criteria found in the course-run enrollability criteria to make the course be unenrollable.
2. Go to the [course enrollment landing page](https://business.sandbox.edx.org/enterprise/2b5a2956-7746-4fc9-9587-bc887ba45973/course/course-v1:edX+DemoX+Demo_Course/enroll/) and notice that you get a message at the top indicating that the course is unavailable for enrollment at this time, and that track selection & the continue button don't show. You'll see something like this:

selection_002

3. Put this course in a program (doesn't matter if all other courses are enrollable), and go to the [program landing page](https://business.sandbox.edx.org/enterprise/2b5a2956-7746-4fc9-9587-bc887ba45973/program/52ad909b-c57d-4ff1-bab3-999813a2479b/enroll/). Notice that you see the message at the top, and also that the purchase/price text, as well as the "continue" button, are hidden. Make sure that the modals work. You'll see something like this:

selection_003
Continuation:
selection_004

**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [ ] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] Called `make static` for webpack bundling if any static content was updated.
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are (reasonably) squashed
- [ ] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=219&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

File: enterprise/templates/enterprise/enterprise_course_enrollment_page.html [hunk 1]
```diff
@@ -41,36 +41,38 @@ <h2 class="course-confirmation-title">{{ confirmation_text }}</h2>
...
-            </div>
-
-            <label for="radio{{ forloop.counter0 }}">
-              <strong class="title">{{ course_mode.title }}</strong>
-              <span class="price">
-                {{ price_text }}:
-                {% if course_mode.final_price and course_mode.original_price != course_mode.final_price %}
-                  <strike>{{ course_mode.original_price }}</strike> {{ course_mode.final_price }}
-                  <div>{{discount_text|safe }}</div>
+        {% if course_enrollable %}
```
Comment (UmanShahzad) [inline enterprise/templates/enterprise/enterprise_course_enrollment_page.html]: This `if` is the only actual addition. The rest was just spacing for this new branch.

File: enterprise/views.py [hunk 2]
```diff
@@ -983,22 +985,22 @@ def get_enterprise_program_enrollment_page(self, request, enterprise_customer, p
...
-        if discount_data.get('total_incl_tax_excl_discounts') is None:
-            add_missing_price_information_message(request, program_title)
-
         one_click_purchase_eligibility = program_details.get('is_learner_eligible_for_one_click_purchase', False)
+        # The following messages shouldn't both appear at the same time, and we prefer the eligibility message.
         if not one_click_purchase_eligibility:
-            add_not_one_click_purchasable_message(request, enterprise_customer, program_title)
+            messages.add_unenrollable_item_message(request, 'program')
+        elif discount_data.get('total_incl_tax_excl_discounts') is None:
+            messages.add_missing_price_information_message(request, program_title)
```
Comment (UmanShahzad) [inline enterprise/views.py]: Ian noted in UI/UX review that we should not have both of these messages be displayed at the same time, as that may confuse the user.

File: enterprise/utils.py [hunk 3]
```diff
@@ -663,3 +666,31 @@ def track_event(user_id, event_name, properties):
                 'clientId': tracking_context.get('client_id')
             }
         })
+
+
+def parse_datetime_generic(datetime_value):
```
Comment (itsjeyd) [inline enterprise/utils.py]: @UmanShahzad Looks like the main purpose of this function is to silence errors that `parse_datetime` might throw. Maybe we can change the name to better reflect that? How about something like `parse_datetime_handle_invalid` or `parse_datetime_allow_invalid`?

File: test_utils/fake_catalog_api.py [hunk 4]
```diff
@@ -45,7 +45,7 @@
         }
     ],
     'start': '2013-02-05T05:00:00Z',
-    'end': '2017-12-31T18:00:00Z',
+    'end': '3000-12-31T18:00:00Z',
```
Comment (itsjeyd) [inline test_utils/fake_catalog_api.py]: @UmanShahzad Wow! The year 3000... It seems so far away that this doesn't even look like a real date to me 😄

File: tests/test_enterprise/views/test_course_enrollment_view.py [hunk 5]
```diff
@@ -131,14 +131,14 @@ def _check_expected_enrollment_page(self, response, expected_context):
...
             'LMS_ROOT_URL': 'http://localhost:8000',
+            'course_enrollable': True,
         }
         default_context.update(expected_context)
         assert response.status_code == 200
         for key, value in default_context.items():
             assert response.context[key] == value  # pylint: disable=no-member
 
     @mock.patch('enterprise.views.render', side_effect=fake_render)
-    @mock.patch('enterprise.views.ProgramDataExtender')
```
Comment (itsjeyd) [inline tests/test_enterprise/views/test_course_enrollment_view.py]: @UmanShahzad This is just a drive-by clean-up, right? It's not related to the main feature that this PR introduces?

File: tests/test_utilities.py [hunk 6]
```diff
@@ -1007,6 +1007,76 @@ def test_track_event_missing_key(self, analytics_mock):
...
+                "enrollment_start": "2014-10-13T13:11:03Z",
+                "enrollment_end": "2999-10-13T13:11:04Z",
+            },
+            False,
+        ),
+        (
+            {
+                "end": "3000-10-13T13:11:01Z",
+                "enrollment_start": "3000-10-13T13:11:03Z",  # enrollment_start > now
+                "enrollment_end": "2999-10-13T13:11:04Z",
```
Comment (itsjeyd) [inline tests/test_utilities.py]: @UmanShahzad Nit: Can you swap the years for `"enrollment_start"` and `"enrollment_end"` around so that they represent a realistic scenario (where the end of the enrollment period doesn't precede the start of the enrollment period)?

File: tests/test_enterprise/views/test_course_enrollment_view.py [hunk 7]
```diff
@@ -266,7 +265,114 @@ def test_get_course_enrollment_page_consent_declined(
...
+            self,
+            enrollable,
+            registry_mock,
+            ecommerce_api_client_mock,
+            enrollment_api_client_mock,
+            course_catalog_client_mock,
+            *args
+    ):  # pylint: disable=unused-argument
+        """
+        The message indicating that the course is currently unopen to new learners is rendered.
```
Comment (itsjeyd) [inline tests/test_enterprise/views/test_course_enrollment_view.py]: @UmanShahzad "unopen" ---> "not open"??? :D

File: enterprise/utils.py (hunk 3 above)
Comment (UmanShahzad) [reply enterprise/utils.py]: I'm going to try, but I think pylint might shut me down. The proposed function names have 2 verbs in them (parse/handle, parse/allow), so that's why I chose something like "_generic" as a prefix, in an attempt to imply that it can handle any type. Maybe `_generic_type` would have been a better prefix. Let's see.

File: test_utils/fake_catalog_api.py (hunk 4 above)
Comment (UmanShahzad) [reply test_utils/fake_catalog_api.py]: ;)

File: tests/test_enterprise/views/test_course_enrollment_view.py (hunk 5 above)
Comment (UmanShahzad) [reply tests/test_enterprise/views/test_course_enrollment_view.py]: Yes, I'm noticed that we could remove this and jumped at the occasion :)

File: tests/test_utilities.py (hunk 6 above)
Comment (UmanShahzad) [reply tests/test_utilities.py]: Yes, sure.

File: enterprise/utils.py (hunk 3 above)
Comment (UmanShahzad) [reply enterprise/utils.py]: Ok, `_handle_invalid` worked as a suffix here, and pylint didn't complain :+1:

File: tests/test_enterprise/views/test_course_enrollment_view.py (hunk 7 above)
Comment (UmanShahzad) [reply tests/test_enterprise/views/test_course_enrollment_view.py]: Haha, well, turns out this exists in the Oxford Dictionary -- https://en.oxforddictionaries.com/definition/unopen

**Comment by UmanShahzad** [comment]
@itsjeyd Thanks for the review! I've pushed an update.

File: tests/test_enterprise/views/test_course_enrollment_view.py (hunk 7 above)
Comment (itsjeyd) [reply tests/test_enterprise/views/test_course_enrollment_view.py]: That's crazy! :D

**Comment by UmanShahzad** [comment]
@itsjeyd Yeah, it looks like someone (or maybe it happened by itself) changed the edX demonstration course's end date to be nothing, so that the course was again enrollable. Try now.

**Comment by itsjeyd** [comment]
@UmanShahzad Huh. I did go into Studio and set the end date for the demo course prior to testing. Will give it another go.

**Comment by UmanShahzad** [comment]
@itsjeyd Yeah but this info is pulled from course-discovery and I think there's a cronjob that updates course-discovery from data in Studio/LMS (don't quote me there :p ) but it didn't run yet.

**Comment by itsjeyd** [comment]
@UmanShahzad Ah, right. And the reason why it worked for the programs landing page is because the corresponding view doesn't call `is_course_run_enrollable`; it uses `is_program_eligible_for_one_click_purchase` info from course-discovery to determine if it should display a message about the program being unenrollable.

**Comment by douglashall** [comment]
FYI, there is no cron job that runs `refresh_course_metadata` on the business sandbox. I did run the command manually on the business sandbox yesterday to pull in some new courses I had created, not sure if the timing there explains what you were seeing. There is a Jenkins job that runs the command in stage and prod.

**Comment by UmanShahzad** [comment]
@douglashall Ah, okay. I think we have a cronjob for this internal to OpenCraft instances only then (as well as edX stage/prod). But yeah, that may have caused it for all we know. Not a big deal.

@itsjeyd I think the reason it showed on the Programs landing page and not the course enrollment one is because the program we're using to test things has [this course run](https://discovery-business.sandbox.edx.org/admin/course_metadata/courserun/3/change/), which has its verified seat type's upgrade deadline in the past (October 13th). This is recent so we didn't notice it before. I'm changing that now to year 3000. :)
